pkg/peers/peers.go

```diff
@@ -90,33 +116,79 @@ func (p *Peers) updatePeers(ctx context.Context) {
 
 	nodes := v1.NodeList{}
 	// get some nodes, but not ourself
-	if err := p.List(readerCtx, &nodes, client.MatchingLabelsSelector{Selector: p.peerSelector}); err != nil {
+	if err := p.List(readerCtx, &nodes, client.MatchingLabelsSelector{Selector: getSelector()}); err != nil {
 		if errors.IsNotFound(err) {
 			// we are the only node at the moment... reset peerList
-			p.peersAddresses = [][]v1.NodeAddress{}
+			p.workerPeersAddresses = [][]v1.NodeAddress{}
 		}
 		p.log.Error(err, "failed to update peer list")
 		return
 	}
+
 	nodesCount := len(nodes.Items)
 	addresses := make([][]v1.NodeAddress, nodesCount)
 	for i, node := range nodes.Items {
 		addresses[i] = node.Status.Addresses
 	}
-	p.peersAddresses = addresses
+	setAddresses(addresses)
 }
 
-func (p *Peers) GetPeersAddresses() [][]v1.NodeAddress {
+func (p *Peers) GetPeersAddresses(role Role) [][]v1.NodeAddress {
 	p.mutex.Lock()
 	defer p.mutex.Unlock()
 
+	var addresses [][]v1.NodeAddress
+	if role == Worker {
+		addresses = p.workerPeersAddresses
+	} else {
+		addresses = p.masterPeersAddresses
+	}
 	//we don't want the caller to be able to change the addresses
 	//so we create a deep copy and return it
-	addressesCopy := make([][]v1.NodeAddress, len(p.peersAddresses))
-	for i := range p.peersAddresses {
-		addressesCopy[i] = make([]v1.NodeAddress, len(p.peersAddresses[i]))
-		copy(addressesCopy, p.peersAddresses)
+	addressesCopy := make([][]v1.NodeAddress, len(addresses))
+	for i := range addressesCopy {
+		addressesCopy[i] = make([]v1.NodeAddress, len(addresses[i]))
+		copy(addressesCopy, addresses)
 	}
 
 	return addressesCopy
 }
+
+func createSelector(hostNameToExclude string, nodeRole Role) labels.Selector {
+	var nodeTypeLabel string
+
+	switch nodeRole {
+	case Master:
+		nodeTypeLabel = GetUsedControlPlaneLabel()
+	default:
+		nodeTypeLabel = WorkerLabelName
+	}
+
+	reqNotMe, _ := labels.NewRequirement(hostnameLabelName, selection.NotEquals, []string{hostNameToExclude})
+	reqPeers, _ := labels.NewRequirement(nodeTypeLabel, selection.Exists, []string{})
+	selector := labels.NewSelector()
+	selector = selector.Add(*reqNotMe, *reqPeers)
+	return selector
+}
+
+func SetControlPlaneLabelType(node *v1.Node) {
```

Comment:
Do we really need all this set/getLabelType  code? Looks a bit overengineered to me, especially because it needs synchronization now because it' s called by 2 places...
Wouldn't a simple util function isControlPlane(node Node) bool good enough?